File: source/c/nested_cleanup.c
// gcc -o nested_cleanup nested_cleanup.c
#include <stdio.h>

void inner(void) {
    printf("inner running\n");
    printf("inner cleanup\n");
}

void middle(void) {
    printf("middle running\n");
    inner();
    printf("middle cleanup\n");
}

void outer(void) {
    printf("outer running\n");
    middle();
    printf("outer cleanup\n");
}

int main(void) {
    outer();
    return 0;
}

File: source/rust/nested_cleanup.rs
struct Cleanup(&'static str);

impl Drop for Cleanup {
    fn drop(&mut self) {
        println!("{} cleanup", self.0);
    }
}

fn inner() {
    let _c = Cleanup("inner");
    println!("inner running");
}

fn middle() {
    let _c = Cleanup("middle");
    println!("middle running");
    inner();
}

fn outer() {
    let _c = Cleanup("outer");
    println!("outer running");
    middle();
}

fn main() {
    outer();
}
